File: Chapter 05/Python/section_5_6_crypto_tracker/after/infrastructure/adapters/coin_gecko_adapter.py
import urllib.request
import json
from ...core.ports.price_provider_port import PriceProviderPort

class CoinGeckoAdapter(PriceProviderPort):
    """THE ADAPTER (Production).
    
    Encapsulates messy HTTP calls and 3rd-party JSON shapes.
    This class is the 'Clarity Engineer' for the CoinGecko API.
    """

    def get_bitcoin_price(self) -> float:
        """Fetches the live Bitcoin price from the CoinGecko REST API.
        
        Returns:
            float: The current price per coin.
        """
        url = "https://api.coingecko.com/api/v3/simple/price?ids=bitcoin&vs_currencies=usd"
        req = urllib.request.Request(url, headers={"User-Agent": "Python Architectural Demo"})

        try:
            with urllib.request.urlopen(req) as response:
                json_data = response.read()
                price_data = json.loads(json_data)
                price = float(price_data["bitcoin"]["usd"])
                
                print("(PROD ADAPTER) Successfully fetched live price.")
                return price
        except Exception as e:
            print(f"(PROD ADAPTER) API Failure: {str(e)}")
            raise

File: Chapter 05/C#/Section_5_6_Crypto_Tracker/After/Infrastructure/Adapters/CoinGeckoAdapter.cs
using System.Collections.Generic;
using System.Net.Http;
using System.Text.Json;
using System.Threading.Tasks;
using Chapter05.CryptoTracker.After.Core.Ports;

namespace Chapter05.CryptoTracker.After.Infrastructure.Adapters
{
    /// <summary>
    /// ADAPTER 2: The Real Production Adapter.
    /// Encapsulates all the messy HTTP calls and 3rd-party JSON shapes here, 
    /// far away from our pure business logic.
    /// </summary>
    public class CoinGeckoAdapter : IPriceProviderPort
    {
        public async Task<decimal> GetBitcoinPrice()
        {
            using var client = new HttpClient();
            client.DefaultRequestHeaders.Add("User-Agent", "C# App");
            
            var json = await client.GetStringAsync("https://api.coingecko.com/api/v3/simple/price?ids=bitcoin&vs_currencies=usd");
            
            // If CoinGecko changes their JSON structure, ONLY this file breaks.
            var priceData = JsonSerializer.Deserialize<Dictionary<string, Dictionary<string, decimal>>>(json);

            // NULL CHECK
            if (priceData == null || !priceData.ContainsKey("bitcoin"))
            {
                throw new Exception("Failed to parse pricing data from the CoinGecko API.");
            }
            return priceData["bitcoin"]["usd"];
        }
    }
}